### DISCUSSION TITLE
Error saving PNG after CAST

### DISCUSSION THREAD

**Discussion by jfconekta** [libvips/libvips#3609] [Q&A] [answered]
## Bug report

**Describe the bug**
Saving a PNG with the function **pngsave** creates always an 8bit image, even when we change the format to **VIPS_FORMAT_USHORT** 

**To Reproduce**
1. Read an image 
2. Change the format of the image using Cast
3. Save the new image
4. The created image using the new version always create a 8bit image

```C++
#include <vips/vips8>

using namespace vips;

int main (int argc, char **argv)
{ 
  if (VIPS_INIT (argv[0])) 
    vips_error_exit (NULL);
  
  VImage out;
  VImage in = VImage::new_from_file (argv[1],
    VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
  
  double avg = in.avg ();
  
  printf ("avg = %g\n", avg);
  printf ("width = %d\n", in.width ());
  
  in = VImage::new_from_file (argv[1],
    VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
  
  out=in;
  out = out.cast(VIPS_FORMAT_USHORT);
  out.pngsave(argv[2], VImage::option()->set("compression",0));
  
  vips_shutdown ();
  
  return (0);
}
```

**Expected behavior**
New image should be 16bit as in previous versions
![image](https://github.com/libvips/libvips/assets/108014638/3991673a-da7e-45b9-adc9-c57e63916e78)

**Actual behavior**
A clear and concise description of what actually happen.

**Screenshots**
![image](https://github.com/libvips/libvips/assets/108014638/43b74876-20fa-4d2a-8144-32d4119f81f1)

**Environment**
 - OS: Ubuntu 22.04
 - Vips: vips-8.12.1 / vips-8.14.3

**Additional context**
With previous versions vips-8.9.1 in ubuntu 20.04  the new image was 16-bits as expected.

**Comment by jcupitt** [comment, marked as answer]
Hi @jfconekta,

We improved the rules for when to write 16-bit PNG in 8.10 (I think it was). You now need to set the image interpretation to RGB16 or GREY16 to write a 16-bit PNG.

In your code, try:

```C++
  out = out.cast(VIPS_FORMAT_USHORT).copy(VImage::option ()->set ("interpretation", "grey16"));
```

Of course the image will look very dark, I expect you know.

**Reply by jfconekta** [reply]
Thanks a lot, is perfectly clear now.
